doc/developer/design/20210811_catalog_only_mode.md

```diff
@@ -0,0 +1,115 @@
+# Catalog-only mode
+
+## Summary
+
+When users have views causing OOM crashes, we want to provide a mechanism that
+still lets users boot into Materialize and take actions to remediate the issue,
+e.g. drop the OOMing view. To accomplish this, we'll introduce a new boot mode
+called "catalog-only mode" that starts Materialize as normal with the exception
+of not shipping any dataflows for user-created objects, such as materialized
+views.
+
+Ideally, we'll also provide a mechanism that lets users enable dataflows for
+their objects while Materialize is running using a statement such as `ALTER
+INDEX <index name> SET ENABLED`.
+
+## Goals
+
+- Support booting in catalog-only mode, which will not:
+    - Ship the dataflow for any user-created object
+    - Support querying user-created objects, such as views
+    - Support inserting into tables
+
+    Any other action should be allowed, such as creating, modifying, or dropping
+    objects.
+- Introduce a `--catalog-only` flag to Materialize that lets users easily turn
+  on catalog-only mode.
+- Add support for `ALTER INDEX ... SET ENABLED` to enable dataflows for objects
+  while Materialilze is running, and lift the restrictions imposed on
+  catalog-only mode for objects that use those dataflows.
+
+  Note that this statement will be "intrinsically idempotent," so `ALTER INDEX
+  ... SET ENABLED` will always succeed, even if Materialize was not booted in
+  catalog only mode or its dataflows are already active. This matches PG's
+  behavior for `ALTER` statements like `ALTER TABLE a ALTER COLUMN x SET NOT
+  NULL`, which you can run repeatedly.
+- Add no more than a single boolean check to the code path of any operation
+  occurring outside of catalog-only mode.
+
+## Non-Goals
+
+- We are not planning to add a statement to re-enable the dataflows for all
+  user-created objects; users can either accomplish this with a script or by
+  simply rebooting _not_ in catalog-only mode.
+
+## Description
+
+The `Coordinator` will handle the enforcing catalog-only mode on objects, i.e.
+disallowing dataflows from shipping, querying user-crated objects, or inserting
+into tables; these are the potentially prohibited actions. To support this enforcement,
+we'll add a new field to the `Coordinator`:
```

Comment:
Can we achieve this with no new variables? We already have code that looks for an index and complains if it doesn't find one (sequence_peek). Tables might need to add a similar check, but that's ok. I think this approach would be easier if we checked for the actual existence of an index instead of looking at something that tells us if an index should be there or not.

Reply:
The challenge here is that all of the places in the code that look for the existence of an index do so by using the ID; at least, the ones I was looking at. If you are aware of others, I'd be very glad to noode on that.

At any rate, to determine the timestamp of the read, the coordinator will look up the index's ID in its `ArrangementFrontiers`, and find that because the dataflow wasn't actually started that the frontier is missing. This prevents the coordinator from even getting as far as sending the request down to the dataflow layer.

This all being said, maybe the better design in this case is to propagate `catalog_only_mode` onto the `ArrangementFrontiers` to teach them that absent values in `catalog_only_mode` means that you've encountered a prohibited operation––I'd need to chat with more folks about the feasibility. The other challenge with that would be forcing some operations on that struct to pay a checking cost, rather than only having to pay the checking cost once per operation.